### ISSUE TITLE
collectd 5.5.0 dns plugin segfaults with kernel 4.1.x (likely a timing issue: plugin init vs. first data reading)

### ISSUE THREAD

**Issue by hnyman** [openwrt/packages#1660]
I get segfault from collectd 5.5.0 if the DNS statistics plugin is enabled. 
ar71xx / WNDR3700

My personal build included collectd 5.5.0 for a few weeks before the #1575 updated it officially. 

* collectd DNS plugin works ok upto trunk build 46391 (July 16th) with kernel 3.18.18.
* collectd DNS plugin segfaults with all later builds with kernel 4.1.x. Earliest segfaulting build that I have is r46441 (Jul 22) with kernel 4.1.3

```
[  104.435977] 
[  104.435977] do_page_fault(): sending SIGSEGV to collectd for invalid write access to 76db3ae8
[  104.444585] epc = 76f6ef30 in dns.so[76f6e000+14000]
[  104.449597] ra  = 0040f084 in collectd[400000+30000]
[  104.454602] 
```
Disabling dns plugin in Luci statistics enables collectd to stay running.

There has been no commits to dns.c at the collectd repo for 9 months, so I see nothing obvious from that angle. Similarly, there are no obviously related changes in Luci during 16-22 July. Thus I highlight the kernel change as my guess as a possible reason.

I am rather uncertain into which direction to go looking for a solution. I found one earlier discussion with references to lua and collectd and a possible solution related to gcc compilation options, so I am including that link as a reference http://stackoverflow.com/questions/12750264/got-sigsegv-while-calling-require-lsqlite3-with-lua-5-1-5 

I might try looking for more exact regression window inside 16-22 July.

**Comment by hnyman** [comment]
I created test builds and:
* r46424 with kernel 3.18.18 works ok. (all feeds reflect the same moment)
* r46432 with kernel 4.1.2 segfaults. (only main svn updates, all feeds the same as with r46424)

so, to me it looks like kernel bump from 3.8 to 4.1 causes this issue.

**Comment by hnyman** [comment]
Just noting my recent observations also here: 

Advice from Collectd developers led me to test the "Readthreads" parameter:
* Openwrt default is "Readthreads 2". Disabling all other plugins, and trying only the DNS plugin with Readthreads 2 makes collectd to crash.
* But decreasing Readthreads to 1 enables the DNS plugin to work. Both as the single read plugin and also with all 10 plugins enabled.

**Comment by hnyman** [comment]
After the realization of the first value reading cycle being made before the plugin initialisation has been completed, I searched for clues and found PR https://github.com/collectd/collectd/pull/1197, which gave me an idea:

If the DNS plugin takes a few more milliseconds to initialise libpcap, then the first read cycle should be slightly postponed in order to reasonably guarantee the completion of the initialisation and the creation of the callback function.

The patch in PR1197 does not guarantee any delay in case the cycle happens to be just exactly at the interval limit, so I modified the idea and crafted my own patch to delay the first plugin read cycle by 5 seconds:

```
--- a/src/daemon/plugin.c
+++ b/src/daemon/plugin.c
@@ -1149,7 +1149,7 @@ static int plugin_insert_read (read_func
 	int status;
 	llentry_t *le;
 
-	rf->rf_next_read = cdtime ();
+	rf->rf_next_read = cdtime () + 5 * (cdtime_t) 1073741824;  // delay first read 5 s
 	rf->rf_effective_interval = rf->rf_interval;
 
 	pthread_mutex_lock (&read_lock);
```
**After applying the patch, I have not been able to make collectd to crash.** 

Still no idea, why kernel bump from 3.8. to 4.1 triggered the problem for me, but apparently some timing slightly changed.
